### PR TITLE
[CI:DOCS]Binding overhauls

### PR DESCRIPTION
**Pull request by baude**
Add binding for networks and begin documentation for binding methods for godoc.  Also, add major functions to their own subpackages so reduce the amount of of method confusion.  So instead of: bindings.ListImages(), we now do a [bindings].images.List().

Also, the connection is passed to each binding method via a context to allow for future growth.

Lastly, add first set of tests.  There are a couple of things to work out for rootless tests yet.

Signed-off-by: Brent Baude <bbaude@redhat.com>

### CODE REVIEW

**Comment by baude** [comment]
please note that although tests are now being added, i have to not added them to CI yet.  I wish to get this PR in first; then sort out the rootless issues.  The turn on CI for them.  Wont be long but this PR is big enough.

**Comment by rhatdan** [comment]
LGTM, except one include problem,

File: pkg/bindings/containers/containers.go [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+	"github.com/containers/libpod/cmd/podman/shared"
+	"github.com/containers/libpod/libpod"
+	"github.com/containers/libpod/pkg/bindings"
+)
+
+// List obtains a list of containers in local storage.  All parameters to this method are optional.
+// The filters are used to determine which containers are listed. The last parameter indicates to only return
+// the most recent number of containers.  The pod and size booleans indicate that pod information and rootfs
+// size information should also be included.  Any finally, the sync bool synchronizes the OCI runtime and
```
Comment (TomSweeneyRedHat) [inline pkg/bindings/containers/containers.go]: nit "Any finally" -> "Finally"

**Comment by baude** [comment]
fixed imports

File: pkg/bindings/connection.go [hunk 2]
```diff
@@ -60,3 +125,24 @@ func (c Connection) newRequest(httpMethod, endpoint string, httpBody io.Reader,
 	response, err := c.client.Do(req) // nolint
```
Comment (jwhonce) [inline pkg/bindings/connection.go]: We should re-initiate the server connection if dropped.  At least once, maybe three times...

File: pkg/bindings/images/images.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+	return deleted, response.Process(nil)
+}
+
+// Tag adds an additional name to locally-stored image.
+func Tag(ctx context.Context, nameOrID string) error {
+	conn, err := bindings.GetConnectionFromContext(ctx)
+	if err != nil {
+		return err
+	}
+	var ()
```
Comment (jwhonce) [inline pkg/bindings/images/images.go]: No variables declared.  No params sent.

File: pkg/bindings/images/images.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	}
+	response, err := conn.NewRequest(http.MethodPost, "/images/prune", nil, params)
+	if err != nil {
+		return deleted, err
+	}
+	return deleted, response.Process(nil)
+}
+
+// Tag adds an additional name to locally-stored image.
+func Tag(ctx context.Context, nameOrID string) error {
```
Comment (jwhonce) [inline pkg/bindings/images/images.go]: Don't you need the tag value here?  Or maybe an array of tags?

File: pkg/bindings/pods/pods.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+func Exists(ctx context.Context, nameOrID string) (bool, error) {
+	conn, err := bindings.GetConnectionFromContext(ctx)
+	if err != nil {
+		return false, err
+	}
+	response, err := conn.NewRequest(http.MethodGet, fmt.Sprintf("/pods/%s/exists", nameOrID), nil, nil)
+	if err != nil {
+		return false, err
+	}
+	return response.StatusCode == http.StatusOK, err
```
Comment (jwhonce) [inline pkg/bindings/pods/pods.go]: See note above on checking status codes.

File: pkg/bindings/test/images_test.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
+package test_bindings
```
Comment (jwhonce) [inline pkg/bindings/test/images_test.go]: Without populating return values from the image functions these cannot be used.

File: pkg/bindings/pods/pods.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+	response, err := conn.NewRequest(http.MethodPost, "/pods/prune", nil, nil)
+	if err != nil {
+		return err
+	}
+	return response.Process(nil)
+}
+
+// List returns all pods in local storage.  The optional filters parameter can
+// be used to refine which pods should be listed.
+func List(ctx context.Context, filters map[string][]string) (*[]libpod.PodInspect, error) {
```
Comment (jwhonce) [inline pkg/bindings/pods/pods.go]: Shouldn't this be []*libpod.PodInspect ?

File: pkg/bindings/containers/containers.go [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+	"github.com/containers/libpod/libpod"
+	"github.com/containers/libpod/pkg/bindings"
+)
+
+// List obtains a list of containers in local storage.  All parameters to this method are optional.
+// The filters are used to determine which containers are listed. The last parameter indicates to only return
+// the most recent number of containers.  The pod and size booleans indicate that pod information and rootfs
+// size information should also be included.  Finally, the sync bool synchronizes the OCI runtime and
+// container state.
+func List(ctx context.Context, filters map[string][]string, last *int, pod, size, sync *bool) ([]shared.PsContainerOutput, error) { // nolint:typecheck
```
Comment (jwhonce) [inline pkg/bindings/containers/containers.go]: This should be []*shared.PsContainerOutput to be consistent with other API calls.

File: pkg/bindings/images/images.go [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+	}
+	if response.StatusCode == http.StatusOK {
+		return true, nil
+	}
+	return false, nil
+}
+
+// List returns a list of images in local storage.  The all boolean and filters parameters are optional
+// ways to alter the image query.
+func List(ctx context.Context, all *bool, filters map[string][]string) ([]handlers.ImageSummary, error) {
```
Comment (jwhonce) [inline pkg/bindings/images/images.go]: This should be []*handlers.ImageSummary to be consistent with other API endpoints

File: pkg/bindings/images/images.go [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+	}
+	return &inspectedData, response.Process(&inspectedData)
+}
+
+func ImageTree(ctx context.Context, nameOrId string) error {
+	return bindings.ErrNotImplemented
+}
+
+// History returns the parent layers of an image.
+func History(ctx context.Context, nameOrID string) ([]handlers.HistoryResponse, error) {
```
Comment (jwhonce) [inline pkg/bindings/images/images.go]: []*handlers.HistoryResponse

File: pkg/bindings/images/images.go [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+	}
+	// TODO this still needs error handling added
+	//_, err := http.Post(c.makeEndpoint("/images/loads"), "application/json", r) //nolint
+	_ = conn
+	return bindings.ErrNotImplemented
+}
+
+// Remove deletes an image from local storage.  The optional force parameter will forcibly remove
+// the image by removing all all containers, including those that are Running, first.
+func Remove(ctx context.Context, nameOrID string, force *bool) ([]map[string]string, error) {
```
Comment (jwhonce) [inline pkg/bindings/images/images.go]: Given this is one image, an array of maps seems wrong.

File: pkg/bindings/images/images.go [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+		params["compress"] = strconv.FormatBool(*compress)
+	}
+	response, err := conn.NewRequest(http.MethodGet, fmt.Sprintf("/images/%s/get", nameOrID), nil, params)
+	if err != nil {
+		return err
+	}
+	if err := response.Process(nil); err != nil {
+		return err
+	}
+	_, err = io.Copy(w, response.Body)
```
Comment (jwhonce) [inline pkg/bindings/images/images.go]: I believe you need a reponse.Body.Close() after the copy.

File: pkg/bindings/network/network.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/containernetworking/cni/libcni"
+	"github.com/containers/libpod/pkg/bindings"
+)
+
+func Create() {}
+func Inspect(ctx context.Context, nameOrID string) (map[string]interface{}, error) {
```
Comment (jwhonce) [inline pkg/bindings/network/network.go]: Why use interface?  Everything has been string-ified to get here.  So you would need to convert them to xxx for this type of return.

File: pkg/bindings/connection.go [hunk 14]
```diff
@@ -17,33 +25,90 @@ type APIResponse struct {
...
 
 func (c Connection) makeEndpoint(u string) string {
-	return fmt.Sprintf("%s%s", defaultConnection, u)
+	// The d character in the url is discarded and is meaningless
+	return fmt.Sprintf("http://d/%s%s", defaultConnectionPath, u)
 }
 
-func (c Connection) newRequest(httpMethod, endpoint string, httpBody io.Reader, params map[string]string) (*APIResponse, error) {
+// NewRequest assembles the http request and returns the response
+func (c Connection) NewRequest(httpMethod, endpoint string, httpBody io.Reader, params map[string]string) (*APIResponse, error) {
```
Comment (jwhonce) [inline pkg/bindings/connection.go]: This should be named DoRequest() or something like that.  It doesn't return a new API Request object().

If this is changed to a variadic function we could hide more boilerplate:

conn.NewRequest(nil, params, "GET", "/images/%s/json", nameOrID)

NewRequest(r io.Reader, p map[string]string, m string, e string, a ...interface{})
  s := fmt.Sprintf(e, ...a)
  // Use s
  .

File: pkg/bindings/images/images.go (hunk 12 above)
Comment (baude) [reply pkg/bindings/images/images.go]: discussed and resolved

File: pkg/bindings/network/network.go (hunk 13 above)
Comment (baude) [reply pkg/bindings/network/network.go]: ah yes, the joys of working with cni ... they are interfaces ...

File: pkg/bindings/images/images.go (hunk 11 above)
Comment (baude) [reply pkg/bindings/images/images.go]: discussed, keep for now

**Comment by baude** [comment]
fixed up with extra sugar @jwhonce

File: pkg/bindings/connection.go [hunk 15]
```diff
@@ -17,46 +25,170 @@ type APIResponse struct {
...
+	var (
+		err            error
+		response       *http.Response
+		safePathValues []string
+	)
+	// Make sure path values are http url safe
+	for _, pv := range pathValues {
+		safePathValues = append(safePathValues, url.QueryEscape(pv))
+	}
+	safeEndpoint := fmt.Sprintf(endpoint, safePathValues)
```
Comment (jwhonce) [inline pkg/bindings/connection.go]: Shouldn't safePathValues be safePathValues[:] ?

Comment (baude) [reply pkg/bindings/connection.go]: @jwhonce the linter puked on that.

**Pull request merged**